Sprint_2

aurum727 opened the pull request:
созданы рекомендованные тест-юниты к классу BooksCollector
- Нельзя добавить одну и ту же книгу дважды.
- Нельзя выставить рейтинг книге, которой нет в списке.
- Нельзя выставить рейтинг меньше 1.
- Нельзя выставить рейтинг больше 10.
- У не добавленной книги нет рейтинга.
- Добавление книги в избранное.
- Нельзя добавить книгу в избранное, если её нет в словаре books_rating.
- Проверка удаления книги из избранного.

collected 9 items
tests.py::TestBooksCollector::test_add_new_book_add_two_books_result_len_books_rating_2 PASSED                                  [ 11%]
tests.py::TestBooksCollector::test_add_new_book_not_add_existing_book_true PASSED                                                         [ 22%]
tests.py::TestBooksCollector::test_get_book_rating_book_not_in_books_rating_none PASSED                                                [ 33%]
tests.py::TestBooksCollector::test_set_book_rating_not_set_rating_less_1_true PASSED                                                           [ 44%]
tests.py::TestBooksCollector::test_set_book_rating_not_set_rating_more_10_true PASSED                                                      [ 55%]
tests.py::TestBooksCollector::test_get_book_rating_book_not_in_dict_none PASSED                                                               [ 66%] 
tests.py::TestBooksCollector::test_add_book_in_favorites_add_book_from_book_ratings_true PASSED                                   [ 77%] 
tests.py::TestBooksCollector::test_add_book_in_favorites_added_book_not_in_book_ratings_true PASSED                             [ 88%] 
tests.py::TestBooksCollector::test_delete_delete_book_from_favorites_deleted_book_not_in_favorites_false PASSED             [100%]

Commit:
new_commit
Sprint_2

kadirovgm commented:
> созданы рекомендованные тест-юниты к классу BooksCollector
> 
> * Нельзя добавить одну и ту же книгу дважды.
> * Нельзя выставить рейтинг книге, которой нет в списке.
> * Нельзя выставить рейтинг меньше 1.
> * Нельзя выставить рейтинг больше 10.
> * У не добавленной книги нет рейтинга.
> * Добавление книги в избранное.
> * Нельзя добавить книгу в избранное, если её нет в словаре books_rating.
> * Проверка удаления книги из избранного.
> 
> collected 9 items tests.py::TestBooksCollector::test_add_new_book_add_two_books_result_len_books_rating_2 PASSED [ 11%] tests.py::TestBooksCollector::test_add_new_book_not_add_existing_book_true PASSED [ 22%] tests.py::TestBooksCollector::test_get_book_rating_book_not_in_books_rating_none PASSED [ 33%] tests.py::TestBooksCollector::test_set_book_rating_not_set_rating_less_1_true PASSED [ 44%] tests.py::TestBooksCollector::test_set_book_rating_not_set_rating_more_10_true PASSED [ 55%] tests.py::TestBooksCollector::test_get_book_rating_book_not_in_dict_none PASSED [ 66%] tests.py::TestBooksCollector::test_add_book_in_favorites_add_book_from_book_ratings_true PASSED [ 77%] tests.py::TestBooksCollector::test_add_book_in_favorites_added_book_not_in_book_ratings_true PASSED [ 88%] tests.py::TestBooksCollector::test_delete_delete_book_from_favorites_deleted_book_not_in_favorites_false PASSED [100%]

Можно лучше: Стоит инфу про тесты положить в файл readme.md

kadirovgm commented on main.py:
```diff
@@ -47,3 +47,20 @@ def delete_book_from_favorites(self, name):
         return self.favorites
+
+
+collector = BooksCollector()
+collector.add_new_book('Гордость и предубеждение и зомби')
+collector.set_book_rating("Три мушкетера", 3)
+print(collector.get_book_rating("Три мушкетера"))
+for i in range(1, 11):
+    print(i)
+
+book = 'Гордость и предубеждение и зомби'
+# обавляем новую книгу book в словарь
+collector.add_new_book(book)
+# добавляем book в список favorites
+collector.add_book_in_favorites(book)
+# удаляем книгу book из списка favorites
+collector.delete_book_from_favorites(book)
+print(collector.get_list_of_favorites_books())
```
Нужно исправить: зачем этот код здесь? в main.py хранятся все методы, которых стоит протестить, не нужно сюда писать какой то выполняемый сценарий, тем более это не в виде теста. Кажется это вообще не нужный код

aurum727 replied:
Да! Сорри! Забыл убрать. Для себя писал.

kadirovgm commented on tests.py:
```diff
@@ -8,17 +9,122 @@ class TestBooksCollector:
-    def test_add_new_book_add_two_books(self):
+
+    # Проверка добавления книг
+    def test_add_new_book_add_two_books_result_len_books_rating_2(self):
```
Нужно исправить: как раз таки название тестов не должны содержать ожидаемый результат, это короткое описание того, что проверяется

kadirovgm commented on tests.py:
```diff
@@ -8,17 +9,122 @@ class TestBooksCollector:
\ No newline at end of file
+    # Нельзя добавить одну и ту же книгу дважды.
+    def test_add_new_book_not_add_existing_book_true(self):
+        # создаем объект BooksCollector м пустым словарем books_rating
```
Нужно исправить: коммент лучше написать один раз одному тесту, краткое описание теста, советую воспользоваться докстрингом (можно загуглить здесь) - это формат комментов, которых чаще всего используют на реальных проектах. 
Обычные комменты # нужны скрее чтобы дописать в конце строки, с новой строки лучше пользоваться докстрингом и только один раз за тест. Если каждую строку комментировать, то на самом деле это мешает восприятию кода

kadirovgm commented on tests.py:
```diff
@@ -8,17 +9,122 @@ class TestBooksCollector:
+        # счетчик для подсчета количества записи тестовой книги в словаре
+        count_test_book_in_books_rating = 0
+        # считаем количество ключей с именем test_book в словаре books_rating
+        for book in collector.get_books_rating().keys():
```
Можно лучше: все верно, но кажется это лишнее, достаточно было проверить, что количество книг равно один после всех манипуляции, но твой тест также проверяет, что книга именно та, что конечно хорошо. Но юнит тесты должны проверять конкретно одну проверку, а именно то, что книг не становится два. Тем более мы знаем, что другие книги не добавляли. Стоит помнить, что мы проверяем белый ящик, когда пишем юнит тесты, и доп проверки тут скорее удлиняют время прохождения тестов.

kadirovgm commented on tests.py:
```diff
@@ -8,17 +9,122 @@ class TestBooksCollector:
+        # пытаемся установить рейтинг 3 для книги не добавленной в словарь
+        collector.set_book_rating(book_not_in_books_rating, 3)
+        # проверяем, что словарь не содержить значение с ключем book_not_in_books_rating
+        assert collector.get_book_rating(book_not_in_books_rating) is None
```
Все верно!

kadirovgm commented on tests.py:
```diff
@@ -8,17 +9,122 @@ class TestBooksCollector:
+        assert collector.get_book_rating(book_not_in_books_rating) is None
+
+    # Нельзя выставить рейтинг меньше 1
+    def test_set_book_rating_not_set_rating_less_1_true(self):
```
Можно лучше: для таких граничных значений можно было бы воспользоваться параметризацией

kadirovgm commented on tests.py:
```diff
@@ -8,17 +9,122 @@ class TestBooksCollector:
+        # устанавливаем рейтинг test_less_1_rating для книги book
+        collector.set_book_rating(book, test_less_1_rating)
+        # проверяем, что рейтинг test_less_1_rating для book не установился
+        assert collector.get_book_rating(book) != test_less_1_rating
```
Нужно исправить: мы знаем, что по умолчанию ожидаем, что метод выставит рейтинг 1, если передать книге рейтинг 0, поэтому тут стоит проверить именно это

kadirovgm commented on tests.py:
```diff
@@ -8,17 +9,122 @@ class TestBooksCollector:
+        # устанавливаем рейтинг test_less_1_rating для книги book
+        collector.set_book_rating(book, test_more_10_rating)
+        # проверяем, что рейтинг test_less_1_rating для book не установился
+        assert collector.get_book_rating(book) != test_more_10_rating
```
Нужно исправить: та же ситуация, что и с тестом выше

kadirovgm commented on tests.py:
```diff
@@ -8,17 +9,122 @@ class TestBooksCollector:
+        # проверяем, что рейтинг test_less_1_rating для book не установился
+        assert collector.get_book_rating(book) != test_more_10_rating
+
+    # проверяем что нельзя присвоить рейтинг книге которой нет в словаре
```
Нужно исправить: не сходится с названием теста. 
_"проверяем что нельзя присвоить рейтинг книге которой нет в словаре"_ - означает, что ты пытаешься присвоить рейтинг книге, а в тесте ты просто проверяешь, что у рандомной книги нет рейтинга, скорее надо попытаться присвоить рейтинг книге методом `set_book_rating`, а потом уже проверять get_book_rating


Но ты эту проверку уже делал в тесте `test_get_book_rating_book_not_in_books_rating_none`

kadirovgm commented on tests.py:
```diff
@@ -8,17 +9,122 @@ class TestBooksCollector:
+        # добавляем новую книгу в список избранного favorites
+        collector.add_book_in_favorites(book)
+        # проеряем вхождение книги book в словарь favorites
+        assert book in collector.get_list_of_favorites_books()
```
верно! молодец :)

kadirovgm commented on tests.py:
```diff
@@ -8,17 +9,122 @@ class TestBooksCollector:
+        # добавляем test_book_not_in_book_rating в список favorites
+        collector.add_book_in_favorites(test_book_not_in_book_rating)
+        # проверям, что добавленная книга test_book_not_in_book_rating отсутствует в списке favorites
+        assert test_book_not_in_book_rating not in collector.get_list_of_favorites_books()
```
Все верно)

kadirovgm commented on tests.py:
```diff
@@ -8,17 +9,122 @@ class TestBooksCollector:
+        # удаляем книгу book из списка favorites
+        collector.delete_book_from_favorites(book)
+        # проверяем, что книга book отсутсвует в списке favorites
+        assert book not in collector.get_list_of_favorites_books()
```
верно)

Commit:
Sprint_2
commit 2
удалена одна лишняя проверка
добавлена фикстура со словарем книг и рейтингов
добвлены три новые проверки

kadirovgm commented on tests.py:
```diff
@@ -11,120 +11,169 @@ class TestBooksCollector:
+        collector.add_new_book(book)
+        rating = 5
+        collector.set_book_rating(book, rating)
+        assert collector.get_book_rating(book) == rating
```
Верно)

kadirovgm commented on tests.py:
```diff
@@ -11,120 +11,169 @@ class TestBooksCollector:
+        for key, value in books_dict.items():
+            collector.add_new_book(key)
+            collector.set_book_rating(key, value)
+        assert len(collector.get_books_with_specific_rating(6)) == 3
```
Верно)
